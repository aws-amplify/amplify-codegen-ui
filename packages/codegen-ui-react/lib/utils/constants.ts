/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
export const COMPOSITE_PRIMARY_KEY_PROP_NAME = 'id';

export const STORAGE_FILE_KEY = 'key';

export const STORAGE_FILE_ALGO_TYPE = 'SHA-1';

export const AMPLIFY_JS_V5 = '5.0.0';

export const AMPLIFY_JS_V6 = '6.0.0';

export const scriptingPatterns = [
  // JavaScript functions
  /eval\s*\(/i,
  /Function\s*\(/i,
  /setTimeout\s*\(/i,
  /setInterval\s*\(/i,
  /new\s+Function/i,
  /import\s*\(/i,
  /require\s*\(/i,

  // DOM manipulation
  /document\./i,
  /window\./i,
  /location\./i,

  // Event handlers
  /on\w+\s*=/i, // matches onerror=, onload=, etc.

  // Dangerous protocols
  /javascript:/i,
  /data:/i,
  /vbscript:/i,

  // Script tags and variations
  /<script/i,
  /<\/script/i,
  /<x:script/i,

  // SVG exploits
  /<svg/i,
  /xlink:href/i,

  // Object prototype attacks
  /\[\s*Symbol\s*\./i,
  /__proto__/i,
  /prototype\s*\./i,

  // Template literal attacks
  /\$\{/i,

  // Base64 indicators
  /base64/i,
];
