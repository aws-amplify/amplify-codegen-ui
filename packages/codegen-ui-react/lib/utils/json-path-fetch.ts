/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { factory, NodeFlags, SyntaxKind, VariableStatement } from 'typescript';

/**
 * does not support array types within objects as it's currently not supported
 *
 * ref: https://stackoverflow.com/questions/45942118/lodash-return-array-of-values-if-the-path-is-valid
 *
 * @param input object input
 * @param path dot notation path for the provided input
 * @param accumlator array
 * @returns returns value at the end of object
 */
export const fetchByPath = <T extends Record<string, any>>(input: T, path: string, accumlator: any[] = []) => {
  const currentPath = path.split('.');
  const head = currentPath.shift();
  if (input && head && input[head] !== undefined) {
    if (!currentPath.length) {
      accumlator.push(input[head]);
    } else {
      fetchByPath(input[head], currentPath.join('.'), accumlator);
    }
  }
  return accumlator[0];
};

/**
 * get the generated output of the fetchByPath function in TS AST
 *
 * @returns VariableStatement
 */
export const getFetchByPathNodeFunction = (): VariableStatement => {
  return factory.createVariableStatement(
    [factory.createModifier(SyntaxKind.ExportKeyword)],
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          factory.createIdentifier('fetchByPath'),
          undefined,
          undefined,
          factory.createArrowFunction(
            undefined,
            [
              factory.createTypeParameterDeclaration(
                factory.createIdentifier('T'),
                factory.createTypeReferenceNode(factory.createIdentifier('Record'), [
                  factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
                  factory.createKeywordTypeNode(SyntaxKind.AnyKeyword),
                ]),
                undefined,
              ),
            ],
            [
              factory.createParameterDeclaration(
                undefined,
                undefined,
                undefined,
                factory.createIdentifier('input'),
                undefined,
                factory.createTypeReferenceNode(factory.createIdentifier('T'), undefined),
                undefined,
              ),
              factory.createParameterDeclaration(
                undefined,
                undefined,
                undefined,
                factory.createIdentifier('path'),
                undefined,
                undefined,
                factory.createStringLiteral(''),
              ),
              factory.createParameterDeclaration(
                undefined,
                undefined,
                undefined,
                factory.createIdentifier('accumlator'),
                undefined,
                factory.createArrayTypeNode(factory.createKeywordTypeNode(SyntaxKind.AnyKeyword)),
                factory.createArrayLiteralExpression([], false),
              ),
            ],
            undefined,
            factory.createToken(SyntaxKind.EqualsGreaterThanToken),
            factory.createBlock(
              [
                factory.createVariableStatement(
                  undefined,
                  factory.createVariableDeclarationList(
                    [
                      factory.createVariableDeclaration(
                        factory.createIdentifier('currentPath'),
                        undefined,
                        undefined,
                        factory.createCallExpression(
                          factory.createPropertyAccessExpression(
                            factory.createIdentifier('path'),
                            factory.createIdentifier('split'),
                          ),
                          undefined,
                          [factory.createStringLiteral('.')],
                        ),
                      ),
                    ],
                    NodeFlags.Const,
                  ),
                ),
                factory.createVariableStatement(
                  undefined,
                  factory.createVariableDeclarationList(
                    [
                      factory.createVariableDeclaration(
                        factory.createIdentifier('head'),
                        undefined,
                        undefined,
                        factory.createCallExpression(
                          factory.createPropertyAccessExpression(
                            factory.createIdentifier('currentPath'),
                            factory.createIdentifier('shift'),
                          ),
                          undefined,
                          [],
                        ),
                      ),
                    ],
                    NodeFlags.Const,
                  ),
                ),
                factory.createIfStatement(
                  factory.createBinaryExpression(
                    factory.createBinaryExpression(
                      factory.createIdentifier('input'),
                      factory.createToken(SyntaxKind.AmpersandAmpersandToken),
                      factory.createIdentifier('head'),
                    ),
                    factory.createToken(SyntaxKind.AmpersandAmpersandToken),
                    factory.createBinaryExpression(
                      factory.createElementAccessExpression(
                        factory.createIdentifier('input'),
                        factory.createIdentifier('head'),
                      ),
                      factory.createToken(SyntaxKind.ExclamationEqualsEqualsToken),
                      factory.createIdentifier('undefined'),
                    ),
                  ),
                  factory.createBlock(
                    [
                      factory.createIfStatement(
                        factory.createPrefixUnaryExpression(
                          SyntaxKind.ExclamationToken,
                          factory.createPropertyAccessExpression(
                            factory.createIdentifier('currentPath'),
                            factory.createIdentifier('length'),
                          ),
                        ),
                        factory.createBlock(
                          [
                            factory.createExpressionStatement(
                              factory.createCallExpression(
                                factory.createPropertyAccessExpression(
                                  factory.createIdentifier('accumlator'),
                                  factory.createIdentifier('push'),
                                ),
                                undefined,
                                [
                                  factory.createElementAccessExpression(
                                    factory.createIdentifier('input'),
                                    factory.createIdentifier('head'),
                                  ),
                                ],
                              ),
                            ),
                          ],
                          true,
                        ),
                        factory.createBlock(
                          [
                            factory.createExpressionStatement(
                              factory.createCallExpression(factory.createIdentifier('fetchByPath'), undefined, [
                                factory.createElementAccessExpression(
                                  factory.createIdentifier('input'),
                                  factory.createIdentifier('head'),
                                ),
                                factory.createCallExpression(
                                  factory.createPropertyAccessExpression(
                                    factory.createIdentifier('currentPath'),
                                    factory.createIdentifier('join'),
                                  ),
                                  undefined,
                                  [factory.createStringLiteral('.')],
                                ),
                                factory.createIdentifier('accumlator'),
                              ]),
                            ),
                          ],
                          true,
                        ),
                      ),
                    ],
                    true,
                  ),
                  undefined,
                ),
                factory.createReturnStatement(
                  factory.createElementAccessExpression(
                    factory.createIdentifier('accumlator'),
                    factory.createNumericLiteral('0'),
                  ),
                ),
              ],
              true,
            ),
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );
};
