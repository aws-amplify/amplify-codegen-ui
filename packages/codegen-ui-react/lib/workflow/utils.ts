/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { ComponentMetadata, InvalidInputError } from '@aws-amplify/codegen-ui';
import Primitive, { PrimitiveChildrenPropMapping } from '../primitive';

export const getChildPropMappingForComponentName = (
  componentMetadata: ComponentMetadata,
  componentName: string,
): string | undefined => {
  const referencedComponentType = componentMetadata.componentNameToTypeMap[componentName];
  if (!referencedComponentType) {
    throw new InvalidInputError(
      `Invalid definition, found reference to component name ${componentName} which wasn't found in the schema.`,
    );
  }
  return PrimitiveChildrenPropMapping[Primitive[referencedComponentType as Primitive]];
};
