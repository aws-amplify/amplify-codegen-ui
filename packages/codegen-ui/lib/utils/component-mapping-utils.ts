/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { StudioComponent, StudioComponentChild } from '../types';

/**
 * Helper to recurse through the component tree and build the name to type mapping.
 */
export const buildComponentNameToTypeMap = (
  component: StudioComponent | StudioComponentChild,
): Record<string, string> => {
  const localMap: Record<string, string> = {};
  if (component.name) {
    localMap[component.name] = component.componentType;
  }
  if (component.children) {
    Object.entries(
      component.children
        .map((child) => buildComponentNameToTypeMap(child))
        .reduce((previous, next) => {
          return { ...previous, ...next };
        }, {}),
    ).forEach(([name, componentType]) => {
      localMap[name] = componentType;
    });
  }
  return localMap;
};
