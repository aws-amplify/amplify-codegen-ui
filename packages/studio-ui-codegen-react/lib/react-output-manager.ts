import { FrameworkOutputManager } from '@amzn/studio-ui-codegen';
import { existsSync, mkdirSync, promises as fs } from 'fs';
import path from 'path';

export class ReactOutputManager extends FrameworkOutputManager<string> {
  async writeComponent(input: string, outputPath: string, componentName: string): Promise<void> {
    console.log('Writing file ', outputPath);

    const { dir } = path.parse(outputPath);

    if (!existsSync(dir)) {
      mkdirSync(dir);
    }

    if (!input) {
      throw new Error('You must call renderComponent before you can save the file.');
    }

    const generatedNotice = `\
/***************************************************************************
 * The contents of this file were generated with Amplify Studio.           *
 * Please refrain from making any modifications to this file.              *
 * Any changes to this file will be overwritten when running amplify pull. *
 **************************************************************************/

`;
    const generatedOutput = `${generatedNotice}${input}`;

    await fs.writeFile(outputPath, generatedOutput);
  }
}
