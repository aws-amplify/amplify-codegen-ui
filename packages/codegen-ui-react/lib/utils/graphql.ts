/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { plural } from 'pluralize';
import { FieldConfigMetadata, InternalError, InvalidInputError } from '@aws-amplify/codegen-ui';
import {
  AwaitExpression,
  CallExpression,
  NodeFlags,
  ObjectLiteralElementLike,
  Statement,
  SyntaxKind,
  factory,
} from 'typescript';
import { ImportCollection } from '../imports';
import { capitalizeFirstLetter, getSetNameIdentifier, lowerCaseFirst } from '../helpers';
import { isBoundProperty, isConcatenatedProperty } from '../react-component-render-helper';
import { Primitive } from '../primitive';
import { DataApiKind, DataStoreRenderConfig, GraphqlRenderConfig, NoApiRenderConfig } from '../react-render-config';
import { isModelDataType } from '../forms/form-renderer-helper/render-checkers';
import { AMPLIFY_JS_V5 } from './constants';
import { getAmplifyJSAPIImport, getAmplifyJSVersionToRender } from '../helpers/amplify-js-versioning';

export enum ActionType {
  CREATE = 'create',
  UPDATE = 'update',
  DELETE = 'delete',
  LIST = 'list',
  GET = 'get',
  GET_BY_RELATIONSHIP = 'getByRelationship',
}

/* istanbul ignore next */
export const isGraphqlConfig = (
  apiConfiguration?: GraphqlRenderConfig | DataStoreRenderConfig | NoApiRenderConfig,
): apiConfiguration is GraphqlRenderConfig => {
  return apiConfiguration?.dataApi === 'GraphQL';
};

export const getGraphqlQueryForModel = (action: ActionType, model: string, byFieldName = ''): string => {
  switch (action) {
    case ActionType.CREATE:
      return `create${model}`;
    case ActionType.UPDATE:
      return `update${model}`;
    case ActionType.DELETE:
      return `delete${model}`;
    case ActionType.GET:
      return `get${model}`;
    case ActionType.LIST:
      return `list${plural(model)}`;
    case ActionType.GET_BY_RELATIONSHIP:
      return `${lowerCaseFirst(plural(model))}By${capitalizeFirstLetter(byFieldName)}`;
    default:
      throw new InvalidInputError(`Action ${action} has no corresponding GraphQL operation`);
  }
};

/* istanbul ignore next */
export const isGraphqlQueryAction = (action: ActionType) => {
  return [ActionType.GET_BY_RELATIONSHIP, ActionType.GET, ActionType.LIST].includes(action);
};

/**
 * Returns a GraphQL call expression and adds to importCollection.
 *
 * @example
 * ```
 * API.graphql({
 *  query: createTodo,
 *  variables: {
 *    input: {
 *      inputs
 *    },
 *  },
 * });
 * ```
 */
/* istanbul ignore next */
export const getGraphqlCallExpression = (
  action: ActionType,
  model: string,
  importCollection: ImportCollection,
  variables?:
    | {
        inputs?: ObjectLiteralElementLike[];
        filters?: ObjectLiteralElementLike[];
      }
    | ObjectLiteralElementLike[],
  byFieldName?: string,
  renderConfigDependencies?: { [key: string]: string },
): CallExpression => {
  const query = getGraphqlQueryForModel(action, model, byFieldName);
  const graphqlVariables: ObjectLiteralElementLike[] = [];
  const graphqlOptions: ObjectLiteralElementLike[] = [
    factory.createPropertyAssignment(
      factory.createIdentifier('query'),
      factory.createCallExpression(
        factory.createPropertyAccessExpression(factory.createIdentifier(query), factory.createIdentifier('replaceAll')),
        undefined,
        [factory.createStringLiteral('__typename'), factory.createStringLiteral('')],
      ),
    ),
  ];

  const mappedImport = getAmplifyJSAPIImport(renderConfigDependencies);
  importCollection.addMappedImport(mappedImport);

  if (isGraphqlQueryAction(action)) {
    importCollection.addGraphqlQueryImport(query);
  } else {
    importCollection.addGraphqlMutationImport(query);
  }

  if (Array.isArray(variables)) {
    graphqlOptions.push(...variables);
  } else {
    if (variables?.inputs) {
      if (isGraphqlQueryAction(action)) {
        variables.inputs.forEach((variable) => {
          graphqlVariables.push(variable);
        });
      } else {
        graphqlVariables.push(
          factory.createPropertyAssignment(
            factory.createIdentifier('input'),
            factory.createObjectLiteralExpression(variables.inputs, true),
          ),
        );
      }
    }
    // filter applies to list
    if ((action === ActionType.LIST || action === ActionType.GET_BY_RELATIONSHIP) && variables?.filters) {
      graphqlVariables.push(
        factory.createPropertyAssignment(
          factory.createIdentifier('filter'),
          factory.createObjectLiteralExpression(variables?.filters, true),
        ),
      );
    }

    if (graphqlVariables.length > 0) {
      graphqlOptions.push(
        factory.createPropertyAssignment(
          factory.createIdentifier('variables'),
          factory.createObjectLiteralExpression(graphqlVariables),
        ),
      );
    }
  }

  importCollection.addModelImport(model);
  const amplifyJSVersion = getAmplifyJSVersionToRender(renderConfigDependencies);
  return factory.createCallExpression(
    factory.createPropertyAccessExpression(
      factory.createIdentifier(amplifyJSVersion === AMPLIFY_JS_V5 ? 'API' : 'client'),
      factory.createIdentifier('graphql'),
    ),
    undefined,
    [factory.createObjectLiteralExpression(graphqlOptions, true)],
  );
};

/* istanbul ignore next */
export const mapFieldArraysToPropertyAssignments = (
  joinTableFields: string[],
  modelFields: string[],
  modelRecord: string,
) => {
  return joinTableFields.map((key, i) => {
    const recordKey = modelFields[i];
    if (!recordKey) {
      throw new InternalError(`Cannot find corresponding key for ${key}`);
    }

    return factory.createPropertyAssignment(
      factory.createIdentifier(key),
      factory.createPropertyAccessExpression(
        factory.createIdentifier(modelRecord),
        factory.createIdentifier(recordKey),
      ),
    );
  });
};

/* istanbul ignore next */
export const getGraphQLJoinTableCreateExpression = (
  relatedJoinTableName: string,
  thisModelRecord: string,
  thisModelPrimaryKeys: string[],
  joinTableThisModelFields: string[],
  relatedModelRecord: string,
  relatedModelPrimaryKeys: string[],
  joinTableRelatedModelFields: string[],
  importCollection: ImportCollection,
  renderConfigDependencies?: { [key: string]: string },
) => {
  const thisModelAssignments = mapFieldArraysToPropertyAssignments(
    joinTableThisModelFields,
    thisModelPrimaryKeys,
    thisModelRecord,
  );
  const relatedModelAssignments = mapFieldArraysToPropertyAssignments(
    joinTableRelatedModelFields,
    relatedModelPrimaryKeys,
    relatedModelRecord,
  );

  return getGraphqlCallExpression(
    ActionType.CREATE,
    relatedJoinTableName,
    importCollection,
    {
      inputs: [...thisModelAssignments, ...relatedModelAssignments],
    },
    undefined,
    renderConfigDependencies,
  );
};

const fetchRelatedRecordEffect = (fetchRecordFunctions: string[]) => {
  return factory.createExpressionStatement(
    factory.createCallExpression(
      factory.createPropertyAccessExpression(factory.createIdentifier('React'), factory.createIdentifier('useEffect')),
      undefined,
      [
        factory.createArrowFunction(
          undefined,
          undefined,
          [],
          undefined,
          factory.createToken(SyntaxKind.EqualsGreaterThanToken),
          factory.createBlock(
            fetchRecordFunctions.map((fetchFunction) => {
              return factory.createExpressionStatement(
                factory.createCallExpression(factory.createIdentifier(fetchFunction), undefined, [
                  factory.createStringLiteral(''),
                ]),
              );
            }),
            true,
          ),
        ),
        factory.createArrayLiteralExpression([], false),
      ],
    ),
  );
};

/* istanbul ignore next */
export const getFetchRelatedRecords = (relatedModelName: string) =>
  `fetch${capitalizeFirstLetter(relatedModelName)}Records`;

/* istanbul ignore next */
export const getFetchRelatedRecordsCallbacks = (
  fieldConfigs: Record<string, FieldConfigMetadata>,
  importCollection: ImportCollection,
  dataApi: DataApiKind = 'DataStore',
  renderConfigDependencies?: { [key: string]: string },
) => {
  const initalFetchRecords: string[] = [];
  const fetchRecordsStatements = Object.entries(fieldConfigs).reduce<Statement[]>(
    (acc, [name, { sanitizedFieldName, relationship, valueMappings, componentType }]) => {
      if (relationship && componentType === Primitive.Autocomplete) {
        const fieldName = name.split('.')[0];
        const renderedFieldName = sanitizedFieldName || fieldName;
        const fetchRecordsVariable = getFetchRelatedRecords(renderedFieldName);
        initalFetchRecords.push(fetchRecordsVariable);
        const setModelLoading = getSetNameIdentifier(`${renderedFieldName}Loading`);
        const setModelFetchOption = getSetNameIdentifier(`${renderedFieldName}Records`);

        const valueConfig = valueMappings?.values[0];
        if (!valueConfig) {
          return acc;
        }
        const displayValueProperty = valueConfig.displayValue || valueConfig.value;

        let displayValues = [];

        if (isConcatenatedProperty(displayValueProperty)) {
          displayValues = displayValueProperty.concat.reduce((displayValue: any, displayValueProp) => {
            if (
              'bindingProperties' in displayValueProp &&
              'field' in displayValueProp.bindingProperties &&
              displayValueProp.bindingProperties.field
            ) {
              displayValue.push(displayValueProp.bindingProperties.field);
            }
            return displayValue;
          }, []);
        } else if (isBoundProperty(displayValueProperty)) {
          displayValues = [displayValueProperty.bindingProperties.field];
        }

        acc.push(
          factory.createVariableStatement(
            undefined,
            factory.createVariableDeclarationList(
              [
                factory.createVariableDeclaration(
                  factory.createIdentifier(fetchRecordsVariable),
                  undefined,
                  undefined,
                  factory.createArrowFunction(
                    [factory.createToken(SyntaxKind.AsyncKeyword)],
                    undefined,
                    [
                      factory.createParameterDeclaration(
                        undefined,
                        undefined,
                        undefined,
                        factory.createIdentifier('value'),
                        undefined,
                        undefined,
                        undefined,
                      ),
                    ],
                    undefined,
                    factory.createToken(SyntaxKind.EqualsGreaterThanToken),
                    factory.createBlock(
                      [
                        factory.createExpressionStatement(
                          factory.createCallExpression(setModelLoading, undefined, [factory.createTrue()]),
                        ),
                        factory.createVariableStatement(
                          undefined,
                          factory.createVariableDeclarationList(
                            [
                              factory.createVariableDeclaration(
                                factory.createIdentifier('newOptions'),
                                undefined,
                                undefined,
                                factory.createArrayLiteralExpression([], false),
                              ),
                            ],
                            NodeFlags.Const,
                          ),
                        ),
                        factory.createVariableStatement(
                          undefined,
                          factory.createVariableDeclarationList(
                            [
                              factory.createVariableDeclaration(
                                factory.createIdentifier('newNext'),
                                undefined,
                                undefined,
                                factory.createStringLiteral(''),
                              ),
                            ],
                            NodeFlags.Let,
                          ),
                        ),
                        factory.createWhileStatement(
                          factory.createBinaryExpression(
                            factory.createBinaryExpression(
                              factory.createPropertyAccessExpression(
                                factory.createIdentifier('newOptions'),
                                factory.createIdentifier('length'),
                              ),
                              factory.createToken(SyntaxKind.LessThanToken),
                              factory.createIdentifier('autocompleteLength'),
                            ),
                            factory.createToken(SyntaxKind.AmpersandAmpersandToken),
                            factory.createParenthesizedExpression(
                              factory.createBinaryExpression(
                                factory.createIdentifier('newNext'),
                                factory.createToken(SyntaxKind.ExclamationEqualsToken),
                                factory.createNull(),
                              ),
                            ),
                          ),
                          factory.createBlock(
                            [
                              factory.createVariableStatement(
                                undefined,
                                factory.createVariableDeclarationList(
                                  [
                                    factory.createVariableDeclaration(
                                      factory.createIdentifier('variables'),
                                      undefined,
                                      undefined,
                                      factory.createObjectLiteralExpression(
                                        [
                                          factory.createPropertyAssignment(
                                            factory.createIdentifier('limit'),
                                            factory.createBinaryExpression(
                                              factory.createIdentifier('autocompleteLength'),
                                              factory.createToken(SyntaxKind.AsteriskToken),
                                              factory.createNumericLiteral('5'),
                                            ),
                                          ),
                                          factory.createPropertyAssignment(
                                            factory.createIdentifier('filter'),
                                            factory.createObjectLiteralExpression(
                                              [
                                                factory.createPropertyAssignment(
                                                  factory.createIdentifier('or'),
                                                  factory.createArrayLiteralExpression(
                                                    displayValues.map((value: string) =>
                                                      factory.createObjectLiteralExpression(
                                                        [
                                                          factory.createPropertyAssignment(
                                                            factory.createIdentifier(value),
                                                            factory.createObjectLiteralExpression(
                                                              [
                                                                factory.createPropertyAssignment(
                                                                  factory.createIdentifier('contains'),
                                                                  factory.createIdentifier('value'),
                                                                ),
                                                              ],
                                                              false,
                                                            ),
                                                          ),
                                                        ],
                                                        false,
                                                      ),
                                                    ),
                                                  ),
                                                ),
                                              ],
                                              false,
                                            ),
                                          ),
                                        ],
                                        true,
                                      ),
                                    ),
                                  ],
                                  NodeFlags.Const,
                                ),
                              ),
                              factory.createIfStatement(
                                factory.createIdentifier('newNext'),
                                factory.createBlock(
                                  [
                                    factory.createExpressionStatement(
                                      factory.createBinaryExpression(
                                        factory.createElementAccessExpression(
                                          factory.createIdentifier('variables'),
                                          factory.createStringLiteral('nextToken'),
                                        ),
                                        factory.createToken(SyntaxKind.EqualsToken),
                                        factory.createIdentifier('newNext'),
                                      ),
                                    ),
                                  ],
                                  true,
                                ),
                                undefined,
                              ),
                              factory.createVariableStatement(
                                undefined,
                                factory.createVariableDeclarationList(
                                  [
                                    factory.createVariableDeclaration(
                                      factory.createIdentifier('result'),
                                      undefined,
                                      undefined,
                                      wrapInParenthesizedExpression(
                                        getGraphqlCallExpression(
                                          ActionType.LIST,
                                          relationship.relatedModelName,
                                          importCollection,
                                          [
                                            factory.createShorthandPropertyAssignment(
                                              factory.createIdentifier('variables'),
                                              undefined,
                                            ),
                                          ],
                                          undefined,
                                          renderConfigDependencies,
                                        ),
                                        [
                                          'data',
                                          getGraphqlQueryForModel(
                                            ActionType.LIST,
                                            capitalizeFirstLetter(relationship.relatedModelName),
                                          ),
                                          'items',
                                        ],
                                      ),
                                    ),
                                  ],
                                  NodeFlags.Const,
                                ),
                              ),
                              factory.createVariableStatement(
                                undefined,
                                factory.createVariableDeclarationList(
                                  [
                                    factory.createVariableDeclaration(
                                      factory.createIdentifier('loaded'),
                                      undefined,
                                      undefined,
                                      factory.createCallExpression(
                                        factory.createPropertyAccessExpression(
                                          factory.createIdentifier('result'),
                                          factory.createIdentifier('filter'),
                                        ),
                                        undefined,
                                        [
                                          factory.createArrowFunction(
                                            undefined,
                                            undefined,
                                            [
                                              factory.createParameterDeclaration(
                                                undefined,
                                                undefined,
                                                undefined,
                                                factory.createIdentifier('item'),
                                                undefined,
                                                undefined,
                                                undefined,
                                              ),
                                            ],
                                            undefined,
                                            factory.createToken(SyntaxKind.EqualsGreaterThanToken),
                                            dataApi === 'GraphQL' && !isModelDataType(fieldConfigs[fieldName])
                                              ? factory.createBinaryExpression(
                                                  factory.createIdentifier(fieldName),
                                                  factory.createToken(SyntaxKind.ExclamationEqualsEqualsToken),
                                                  factory.createPropertyAccessExpression(
                                                    factory.createIdentifier('item'),
                                                    factory.createIdentifier('id'),
                                                  ),
                                                )
                                              : factory.createPrefixUnaryExpression(
                                                  SyntaxKind.ExclamationToken,
                                                  factory.createCallExpression(
                                                    factory.createPropertyAccessExpression(
                                                      factory.createIdentifier(`${fieldName}IdSet`),
                                                      factory.createIdentifier('has'),
                                                    ),
                                                    undefined,
                                                    [
                                                      factory.createCallChain(
                                                        factory.createPropertyAccessExpression(
                                                          factory.createIdentifier('getIDValue'),
                                                          factory.createIdentifier(fieldName),
                                                        ),
                                                        factory.createToken(SyntaxKind.QuestionDotToken),
                                                        undefined,
                                                        [factory.createIdentifier('item')],
                                                      ),
                                                    ],
                                                  ),
                                                ),
                                          ),
                                        ],
                                      ),
                                    ),
                                  ],
                                  NodeFlags.AwaitContext,
                                ),
                              ),
                              factory.createExpressionStatement(
                                factory.createCallExpression(
                                  factory.createPropertyAccessExpression(
                                    factory.createIdentifier('newOptions'),
                                    factory.createIdentifier('push'),
                                  ),
                                  undefined,
                                  [factory.createSpreadElement(factory.createIdentifier('loaded'))],
                                ),
                              ),
                              factory.createExpressionStatement(
                                factory.createBinaryExpression(
                                  factory.createIdentifier('newNext'),
                                  factory.createToken(SyntaxKind.EqualsToken),
                                  factory.createPropertyAccessExpression(
                                    factory.createIdentifier('result'),
                                    factory.createIdentifier('nextToken'),
                                  ),
                                ),
                              ),
                            ],
                            true,
                          ),
                        ),
                        factory.createExpressionStatement(
                          factory.createCallExpression(setModelFetchOption, undefined, [
                            factory.createCallExpression(
                              factory.createPropertyAccessExpression(
                                factory.createIdentifier('newOptions'),
                                factory.createIdentifier('slice'),
                              ),
                              undefined,
                              [factory.createNumericLiteral('0'), factory.createIdentifier('autocompleteLength')],
                            ),
                          ]),
                        ),
                        factory.createExpressionStatement(
                          factory.createCallExpression(setModelLoading, undefined, [factory.createFalse()]),
                        ),
                      ],
                      true,
                    ),
                  ),
                ),
              ],
              NodeFlags.Const,
            ),
          ),
        );
      }

      return acc;
    },
    [],
  );
  return [...fetchRecordsStatements, fetchRelatedRecordEffect(initalFetchRecords)];
};

export function wrapInParenthesizedExpression(callExpression: CallExpression, accessors: string[]): AwaitExpression {
  return accessors.reduce(
    (acc: any, _, index: any, initialValue: string[]) =>
      factory.createPropertyAccessChain(
        acc,
        factory.createToken(SyntaxKind.QuestionDotToken),
        factory.createIdentifier(initialValue[index]),
      ),
    factory.createAwaitExpression(callExpression),
  );
}
