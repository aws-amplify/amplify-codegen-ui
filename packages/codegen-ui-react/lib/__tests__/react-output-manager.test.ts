/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import { mocked } from 'ts-jest/utils'; // eslint-disable-line import/no-extraneous-dependencies
import { existsSync, mkdirSync, promises as fs } from 'fs';
import { ReactOutputManager } from '../react-output-manager';

jest.mock('fs', () => ({
  promises: {
    writeFile: jest.fn(),
  },
  existsSync: jest.fn().mockReturnValue(true),
  mkdirSync: jest.fn(),
}));

describe('react-output-manager', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('writeComponent', () => {
    const outputPath = 'output/MyComponent.js';
    const input = 'My Component Text';

    test('writes component', async () => {
      await new ReactOutputManager().writeComponent(input, outputPath);

      expect(mocked(fs.writeFile).mock.calls).toMatchSnapshot();
    });

    test('writes component if dir does not exist', async () => {
      mocked(existsSync).mockReturnValue(false);

      await new ReactOutputManager().writeComponent(input, outputPath);

      expect(mocked(mkdirSync)).toHaveBeenCalledWith('output');
      expect(mocked(fs.writeFile)).toHaveBeenCalled();
    });

    test('throws error if component text is not set', async () => {
      expect.assertions(1);
      await expect(new ReactOutputManager().writeComponent('', outputPath)).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
