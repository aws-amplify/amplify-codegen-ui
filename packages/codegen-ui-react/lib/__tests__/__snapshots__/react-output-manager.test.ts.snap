// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-output-manager writeComponent throws error if component text is not set 1`] = `"You must call renderComponent before you can save the file."`;

exports[`react-output-manager writeComponent writes component 1`] = `
Array [
  Array [
    "output/MyComponent.js",
    "/***************************************************************************
 * The contents of this file were generated with Amplify Studio.           *
 * Please refrain from making any modifications to this file.              *
 * Any changes to this file will be overwritten when running amplify pull. *
 **************************************************************************/

My Component Text",
  ],
]
`;
